src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11KeyWrapCipher.java

```diff
@@ -0,0 +1,816 @@
+                break;
+                case Cipher.DECRYPT_MODE:
+                    token.p11.C_DecryptInit(session.id(), mechWithParams,
+                            p11KeyID);
+                break;
+                }
+            } catch (PKCS11Exception e) {
+                session = token.releaseSession(session);
+                throw e;
+            } finally {
+                p11Key.releaseKeyID();
+            }
+        }
+        initialized = true;
+    }
+
+    // if doFinal(inLen) is called, how big does the output buffer have to be?
+    private int doFinalLength(int inLen) {
+        if (inLen < 0) {
+            throw new ProviderException("Invalid negative input length");
+        }
+
+        int result = inLen + dataBuffer.size();
+        boolean encrypt = (opmode == Cipher.ENCRYPT_MODE ||
+            opmode == Cipher.WRAP_MODE);
+        if (encrypt) {
+            if (type == KeyWrapType.KW_PKCS5PADDING) {
+                // add potential pad length, i.e. 1-8
+                result += (BLK_SIZE - (result & (BLK_SIZE - 1)));
+            } else if (type == KeyWrapType.KWP_NOPADDING &&
+                    (result & (BLK_SIZE - 1)) != 0) {
+                // add potential pad length, i.e. 0-7
+                result += (BLK_SIZE - (result & (BLK_SIZE - 1)));
+            }
+            result += BLK_SIZE; // add the leading block including the ICV
+        } else {
+            result -= BLK_SIZE; // minus the leading block including the ICV
+        }
+        return (result > 0? result : 0);
+    }
+
+    // reset the states to the pre-initialized values
+    // set initialized to false, cancel operation, release session, and
+    // reset dataBuffer
+    private void reset(boolean doCancel) {
+        if (!initialized) {
+            return;
+        }
+        initialized = false;
+
+        try {
+            if (session == null) {
+                return;
+            }
+
+            if (doCancel && token.explicitCancel) {
+                cancelOperation();
+            }
+        } finally {
+            session = token.releaseSession(session);
+            dataBuffer.reset();
+        }
+    }
+
+    // see JCE spec
+    protected byte[] engineUpdate(byte[] in, int inOfs, int inLen) {
+        int n = implUpdate(in, inOfs, inLen);
+        return new byte[0];
+    }
+
+    // see JCE spec
+    protected int engineUpdate(byte[] in, int inOfs, int inLen, byte[] out,
+            int outOfs) throws ShortBufferException {
+        implUpdate(in, inOfs, inLen);
+        return 0;
+    }
+
+    // see JCE spec
+    @Override
+    protected int engineUpdate(ByteBuffer inBuffer, ByteBuffer outBuffer)
+            throws ShortBufferException {
+        implUpdate(inBuffer);
+        return 0;
+    }
+
+    // see JCE spec
+    protected byte[] engineDoFinal(byte[] in, int inOfs, int inLen)
+            throws IllegalBlockSizeException, BadPaddingException {
+        int minOutLen = doFinalLength(inLen);
```

Comment:
nit: seems like this could be maxOutLen given it's the length used to allocate out[].  It can't be any larger, otherwise the operations would fail

Reply:
Sure, I will rename it to maxOutLen